### PR TITLE
Making min/max temperatures for Material curves more discoverable

### PR DESCRIPTION
**Pull request by john-science**
## Description

The [problem people were having](https://github.com/terrapower/armi/issues/715), that this PR attempts to address, is that in our material definitions, we have a _lot_ of hidden numbers, for instance, the min/max values that a temperature-dependent curve is valid over:

https://github.com/terrapower/armi/blob/cec12c998a9579f0c21fd2fbd1f863767a06fa52/armi/materials/ht9.py#L60-L68

There was already a solution for this for half-a-dozen materials in ARMI:

https://github.com/terrapower/armi/blob/cec12c998a9579f0c21fd2fbd1f863767a06fa52/armi/materials/siC.py#L68-L73

And:

https://github.com/terrapower/armi/blob/cec12c998a9579f0c21fd2fbd1f863767a06fa52/armi/materials/siC.py#L85-L89

This PR just extends this solution to all ARMI materials with temperature-dependent curves. The goal here is to solve the complaint/request by making ARMI materials:

1. More discoverable
2. More consistent

---

## Checklist

- [x] The code is understandable and maintainable to people beyond the author.
- [x] Tests have been added/updated to verify that the new or changed code works.
- [x] There is no commented out code in this PR.
- [x] The commit message follows [good practices](https://terrapower.github.io/armi/developer/tooling.html).
- [x] All docstrings are still up-to-date with these changes.
- [x] Documentation added/updated in the `doc` folder.

### CODE REVIEW

**Comment by john-science** [comment]
@[alexhjames](https://github.com/alexhjames) Just FYI.

File: armi/materials/material.py [hunk 1]
```diff
@@ -45,26 +45,31 @@ class Material(composites.Leaf):
     A material is made up of elements or isotopes. It has bulk properties like mass density.
     """
 
+    # State parameter definitions
     pDefs = materialParameters.getMaterialParameterDefinitions()
-    """State parameter definitions"""
 
+    # Indication of where the material is loaded from (may be plugin name)
     DATA_SOURCE = "ARMI"
```
Comment (ntouran) [inline armi/materials/material.py:52]: this breaks sphinx rendering. Sphinx treats those docstrings as special and renders them nicely. see built docs. Undo or justify. 

see for example https://terrapower.github.io/armi/.apidocs/armi.materials.material.html#armi.materials.material.Material.DATA_SOURCE

**Review by ntouran — CHANGES_REQUESTED** [review]
very impressive! Just a few requests

File: armi/materials/alloy200.py [hunk 2]
```diff
@@ -99,7 +101,9 @@ def linearExpansion(self, Tk=None, Tc=None):
             instantaneous coefficient of thermal expansion of Alloy 200 (1/C)
         """
         Tk = getTk(Tc, Tk)
-        self.checkTempRange(73.15, 1273.15, Tk, "linear expansion")
+        (Tmin, Tmax) = self.propertyValidTemperature["linear expansion"][0]
+        self.checkTempRange(Tmin, Tmax, Tk, "linear expansion")
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/alloy200.py:105]: I suppose that's cleaner.

I was trying to put the onus on the Material dev to make sure the units lined up.

Looking at it with fresh eyes, the dev has to ensure those units line up anyway. Why not help them out by writing less code?  

Sure, I'll give it a whirl and see.  Thanks!

File: armi/materials/cs.py [hunk 3]
```diff
@@ -12,14 +12,30 @@
...
+        """
+        https://en.wikipedia.org/wiki/Caesium
+        """
+        Tk = getTk(Tc, Tk)
+        if Tk < self.meltingPoint():
+            return 1.93  # g/cm3
+        else:
+            return 1.843  # g/cm3
+
+    def meltingPoint(self):
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/cs.py:40]: Ah!

The standard I found in ARMI was just `meltingPoint`.  If you are okay with me kind of throwing that change into this PR as well, I would _love_ too.

File: armi/materials/inconel600.py [hunk 4]
```diff
@@ -17,12 +17,18 @@
...
 
-from armi.utils.units import getTc
 from armi.materials.material import Material
+from armi.utils.units import getTc
 
 
 class Inconel600(Material):
     name = "Inconel600"
+    propertyValidTemperature = {
+        "heat capacity": ((20, 900), "C"),
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/inconel600.py:27]: True.

Also, I don't love `propertyValidTemperature` as a name.  

BUT, we have downstream projects that use this name in their own materials library, so I am trying to preserve their names and data structures, to make their lives easier.

File: armi/materials/material.py (hunk 1 above)
Comment (john-science) [reply armi/materials/material.py:52]: Eww. I don't like that at all.

But sure, I'll fix it back. Thanks!

File: armi/materials/custom.py [hunk 5]
```diff
@@ -31,7 +31,6 @@ class Custom(Material):
 
     name = "Custom Material"
     enrichedNuclide = "U235"
-    correctDensityAfterApplyInputParams = False
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/custom.py:34]: It does not exist in any Nala or Pylot repo.  I checked before removing it.

[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/custom.py:34]: I checked all downstream repos I have access to, and that is not used anywhere.

File: armi/materials/cs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/cs.py:40]: Chris and I were talking about other such consistency improvements that we should do, so I opened a GitHub Discussion for them here:  https://github.com/terrapower/armi/discussions/767

**Comment by john-science** [comment]
As a bonus, this PR fixes the following issues I found while working on this PR:

* BUG: a few materials were using `runLog` without importing it.
* BUG/TYPO: Several materials were passing the wrong string to `checkTempRange()`.
* TYPO: Several materials were misspelling "expansion" as "expansionn".
* BUG?: Californium had no data at all before.
* BUG?: Cesium had no data at all before.

**Review by opotowsky — APPROVED** [review]
Late to the game but figured I'd add an approving review anyway. Nice work!

File: armi/materials/mgO.py [hunk 6]
```diff
@@ -22,12 +22,28 @@
...
 
+    def density(self, Tk=None, Tc=None):
+        """same reference as linear expansion. Table II.
+        Reference density is from Wolfram Alpha At STP (273 K)"""
+        Tk = getTk(Tc, Tk)
+        rho0 = 3.58
+        self.checkPropertyTempRange("density", Tk)
+        dLL = self.linearExpansionPercent(Tk=Tk)
+
+        dRho = (1 - (1 + dLL) ** 3) / (1 + dLL) ** 3
```
Comment (keckler) [inline armi/materials/mgO.py:43]: I understand this is not the fault of this PR, but I'm noticing here that we have a number of density methods which return the expanded density in 3D. Here, for SiC, and I think other materials as well.

This is directly in contradiction with the base method docstring, which says the `density()` method returns the density as expanded in 2D: https://github.com/terrapower/armi/blob/main/armi/materials/material.py#L337

I think we probably need to escalate this issue to somewhere where we can track this for all of the material classes in the framework.

Comment (keckler) [reply armi/materials/mgO.py:43]: This appears to have been resolved in #868 .

File: armi/materials/material.py [hunk 7]
```diff
@@ -539,25 +539,44 @@ def removeLumpedFissionProducts(self) -> None:
...
+
+        val : float
+            The value to check whether it is between minT and maxT.
+
+        Notes
+        -----
+        This was designed as a convience method for ``checkTempRange``.
+        """
+        (minT, maxT) = self.propertyValidTemperature[label][0]
+        self.checkTempRange(minT, maxT, val, label)
```
Comment (keckler) [inline armi/materials/material.py:559]: @john-science Sorry for the super late comment on this PR...

I noticed that the standard syntax for entering into the `propertyValidTemperature` dict includes the units for the range.

However I also noticed that the above method doesn't do anything with those units.

Was it the intention to do some internal conversion on the units here to match whatever was passed in?

**Pull request merged**